File: src/assembler.py
"""
assembler.py

ページごとの Markdown を連結し、最終的な final.md を生成するモジュール。
"""

import logging
import os
import re

logger = logging.getLogger(__name__)


def _fix_unclosed_code_blocks(page_md: str, page_num: int) -> str:
    """閉じられていないコードブロックを検出し、末尾に閉じタグを追加する。"""
    open_count = 0
    close_count = 0
    for line in page_md.splitlines():
        stripped = line.strip()
        if stripped.startswith("```") and len(stripped) > 3:
            open_count += 1
        elif stripped == "```":
            close_count += 1
    if open_count > close_count:
        missing = open_count - close_count
        logger.warning(f"Page {page_num}: {missing}個の閉じられていないコードブロックを検出。自動で閉じます。")
        page_md += "\n```" * missing
    return page_md


def assemble_final_md(page_md_paths, run_output_dir, images_dir, pdf_filename):
    """
    ページMDを番号順に連結し、2種類のMDファイルを生成する。
    - 画像付きバージョン: ページ先頭に元のページ画像リンクを挿入
    - ピュアMarkdownバージョン: テキストのみ

    Args:
        page_md_paths (list[str]): ページMDファイルパスのリスト(番号順)
        run_output_dir (str): 出力ディレクトリ(output/<timestamp>/)
        images_dir (str): 画像ディレクトリ
        pdf_filename (str): 元のPDFファイル名

    Returns:
        tuple[str, str]: (画像付きMDファイルパス, ピュアMDファイルパス)
    """
    pages_with_images = []
    pages_pure = []
    for i, md_path in enumerate(page_md_paths, 1):
        with open(md_path, encoding="utf-8") as f:
            page_md = f.read()
        # 行番号除去
        page_md = re.sub(r"^L\d{3}:\s?", "", page_md, flags=re.MULTILINE)
        # 閉じられていないコードブロックを修正
        page_md = _fix_unclosed_code_blocks(page_md, i)
        pages_pure.append(f"<!-- page {i} -->\n\n" + page_md.strip())
        # ページ先頭に画像リンクを挿入したバージョン
        image_name = f"page_{str(i).zfill(3)}.png"
        image_rel_path = os.path.relpath(os.path.join(images_dir, image_name), run_output_dir)
        page_with_image = f"![page {i}]({image_rel_path})\n\n" + page_md
        pages_with_images.append(page_with_image.strip())

    basename = os.path.splitext(os.path.basename(pdf_filename))[0]

    # 画像付きバージョン
    final_with_images = "\n\n---\n\n".join(pages_with_images)
    with_images_path = os.path.join(run_output_dir, f"{basename}.md")
    with open(with_images_path, "w", encoding="utf-8") as f:
        f.write(final_with_images)

    # ピュアMarkdownバージョン
    final_pure = "\n\n---\n\n".join(pages_pure)
    pure_path = os.path.join(run_output_dir, f"{basename}_pure.md")
    with open(pure_path, "w", encoding="utf-8") as f:
        f.write(final_pure)

    return with_images_path, pure_path

File: src/assembler.ts
import { copyFileSync, existsSync, mkdirSync, readdirSync, readFileSync, writeFileSync } from "node:fs";
import path from "node:path";

/**
 * 閉じられていないコードブロックを検出し、末尾に閉じタグを追加する。
 */
function fixUnclosedCodeBlocks(pageMd: string, pageNum: number): string {
	let openCount = 0;
	let closeCount = 0;
	for (const line of pageMd.split("\n")) {
		const stripped = line.trim();
		if (stripped.startsWith("```") && stripped.length > 3) {
			openCount++;
		} else if (stripped === "```") {
			closeCount++;
		}
	}
	if (openCount > closeCount) {
		const missing = openCount - closeCount;
		console.warn(`Page ${pageNum}: ${missing}個の閉じられていないコードブロックを検出。自動で閉じます。`);
		pageMd += "\n```".repeat(missing);
	}
	return pageMd;
}

/**
 * 同名ファイルを避けて連番(_2, _3, ...)を付与したパスを返す。
 */
function uniquePath(dir: string, basename: string, suffix: string): string {
	let candidate = path.join(dir, `${basename}${suffix}`);
	if (!existsSync(candidate)) return candidate;
	let i = 2;
	while (true) {
		candidate = path.join(dir, `${basename}_${i}${suffix}`);
		if (!existsSync(candidate)) return candidate;
		i++;
	}
}

/**
 * 同名dirを避けて連番付与したパスを返す。
 */
function uniqueDirPath(dir: string, name: string): string {
	let candidate = path.join(dir, name);
	if (!existsSync(candidate)) return candidate;
	let i = 2;
	while (true) {
		candidate = path.join(dir, `${name}_${i}`);
		if (!existsSync(candidate)) return candidate;
		i++;
	}
}

export interface AssembleOptions {
	pageMdPaths: string[];
	imagesDir: string;
	pdfPath: string;
	withImages: boolean;
}

export interface AssembleResult {
	purePath: string;
	withImagesPath: string | null;
}

/**
 * ページMDを番号順に連結し、入力PDFと同じdirへ書き出す。
 * - デフォルト: pure MD のみ (`<basename>.md`)
 * - withImages 指定時: 加えて `<basename>.with-images.md` と `<basename>_images/` を出力
 */
export function assembleFinalMd(opts: AssembleOptions): AssembleResult {
	const { pageMdPaths, imagesDir, pdfPath, withImages } = opts;

	const outputDir = path.dirname(path.resolve(pdfPath));
	const basename = path.basename(pdfPath, path.extname(pdfPath));

	const pagesPure: string[] = [];
	const pagesWithImages: string[] = [];

	// withImages の場合は画像コピー先 dir を準備
	let imagesOutDir: string | null = null;
	let imagesOutDirName: string | null = null;
	if (withImages) {
		imagesOutDir = uniqueDirPath(outputDir, `${basename}_images`);
		imagesOutDirName = path.basename(imagesOutDir);
		mkdirSync(imagesOutDir, { recursive: true });
	}

	for (let i = 0; i < pageMdPaths.length; i++) {
		const mdPath = pageMdPaths[i];
		const pageNum = i + 1;
		let pageMd = readFileSync(mdPath, "utf-8");

		// 行番号除去
		pageMd = pageMd.replace(/^L\d{3}:\s?/gm, "");
		// 閉じられていないコードブロックを修正
		pageMd = fixUnclosedCodeBlocks(pageMd, pageNum);

		pagesPure.push(`<!-- page ${pageNum} -->\n\n${pageMd.trim()}`);

		if (withImages && imagesOutDirName) {
			const imageName = `page_${String(pageNum).padStart(3, "0")}.png`;
			const imageRelPath = `${imagesOutDirName}/${imageName}`;
			pagesWithImages.push(`![page ${pageNum}](${imageRelPath})\n\n${pageMd}`.trim());
		}
	}

	// 画像コピー
	if (withImages && imagesOutDir) {
		for (const file of readdirSync(imagesDir)) {
			copyFileSync(path.join(imagesDir, file), path.join(imagesOutDir, file));
		}
	}

	// pure MD 書き出し
	const purePath = uniquePath(outputDir, basename, ".md");
	writeFileSync(purePath, pagesPure.join("\n\n---\n\n"), "utf-8");

	let withImagesPath: string | null = null;
	if (withImages) {
		withImagesPath = uniquePath(outputDir, basename, ".with-images.md");
		writeFileSync(withImagesPath, pagesWithImages.join("\n\n---\n\n"), "utf-8");
	}

	return { purePath, withImagesPath };
}
